temp paste headers fix + heading spacing

charles01pd2018 opened the pull request:
## Pasting headers from own editor works

1. I saw the line of code that parsed through the HTML before the paste event (PasteHandler.ts) and commented that out. Check the file to see the explanation.
2. I noticed there were a lot of typescript bugs so I just corrected a few in files that I came across
3. Added type-check script -> goes through the entire project and checks for typescript errors

charles01pd2018 committed:
temp paste headers fix

charles01pd2018 committed:
heading line-height

hadielyakhni commented on package.json:
```diff
@@ -11,7 +11,8 @@
     "preversion": "npm run build",
-    "build-storybook": "build-storybook"
+    "build-storybook": "build-storybook",
+    "type-check": "node_modules/.bin/tsc --noEmit"
```
for consistency with other scripts, we can omit the `node_modules/.bin` prefix, and let the package manager resolve it for us!

```suggestion
    "type-check": "tsc --noEmit"
```

lmk If i missed something!!

-------

btw, talking about package managers, this repo used to use `yarn` then we switched to `npm` and maybe we switched back to `yarn` - i forgot tbh :/ so you'll notice we have npm scripts in the package.json, like `"preversion": "npm run build"`, at the same time we have `yarn.lock` instead of `package-lock.json`, and we have `resolutions` in package.json, which only works with yarn (npm uses `overrides`) - so you get it 😄 we're mixing yarn and npm! so we'll need to fix that inconsistency; if you can do some research and choose one of them! (ps: pnpm is another option, tried it before, but it caused some problems when using this package in the main web app)

charles01pd2018 replied:
got it ok thanks for the context ill make sure to look into npm and yarn

hadielyakhni replied:
awesome thanks! please note that yarn has 2 versions, [classic (v1)](https://github.com/yarnpkg/yarn) and a [new version (v2)](https://github.com/yarnpkg/berry) - we use v1 in the main app, as v2 is fundamentally different from yarn v1 and npm, and it's not too mature yet i think

hadielyakhni commented on src/plugins/PasteHandler.ts:
```diff
@@ -34,11 +34,19 @@ export default class PasteHandler extends Extension {
-            return replaceHeaderByStrong(html);
-          },
+          /**
+           * I'm not sure the use case for this previously
+           * from what I understand this pasted any header tags (copied from within the editor or outside)
+           * to just be <strong>. Without this line of code, header pasted copied from the editor work
+           * HOWEVER, headers copied from other sites would be pasted as regular text
```
so previously, before this change was made, copying notes from editor to editor worked fine, including pasting headers. However, when copying content with headers from another website, it worked, until we refresh the page! when refreshing, I think an error was thrown - I will test quickly though, and attach a recording of the broken behavior!

charles01pd2018 replied:
ok got it got it -> i knew that piece of code was there for a reason but couldn't find the use case for it haha. The screen recordings helped a lot got it thanks so much - I'll keep working on this

charles01pd2018 replied:
Ok, I think I found the issue but idrk what the best method to solve it is.

Basically, the reason why heading pastes aren't working from certain websites is that our specific editor schema is expecting an inner ```<span class='heading-content'>Inner Content</span> ``` for each heading element, but some websites just directly render text content as innerHTML. That's why we are getting the ```firstChild``` bug, because we expect a child to exist within heading elements.

So I think the best move is to parse the HTML and append ```span``` tags with each respective heading tag IF they aren't there already. Idk how to do this lol.

EDIT: @hadielyakhni if you have experience with the [Dom Parser API](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) lmk! If not I'll see if I can come up with a solution but it might take some time.

hadielyakhni replied:
oh, great finding! i am not too familiar with the dom parser API, i can't think of a good way to add spans if they're missing :) though, do you think we can just do 🙂 

```js
// contentElement: ".heading-content"
```

hadielyakhni commented:
@charles01pd2018, here is the issue, i copied all the content from [this page](https://blog.openreplay.com/javascript-seo-best-practices/) (cmd+a, cmd+c):

on the main branch (headers are transformed to strong), pasting works, but the note brokes after refreshing:
https://streamable.com/mbxydm

on this PR's branch, pasting didn't work:
https://streamable.com/9gp4ge

charles01pd2018 committed:
html check + heading paste

HTML CHECK
Now, the html check will confirm that IF the html content is in a code block then it will be ignored

Heading Paste
Remove contentElement + check for em fixes heading paste

charles01pd2018 committed:
package.json

only yarn

charles01pd2018 committed:
fix isHTML return

charles01pd2018 commented:
@Apandoh this one needs a little more testing - I think this works but can you confirm with some older saved data in HTML format?

hadielyakhni commented on src/plugins/PasteHandler.ts:
```diff
@@ -140,11 +140,14 @@ export default class PasteHandler extends Extension {
-              const transaction = view.state.tr.replaceSelection(slice);
-              view.dispatch(transaction);
-              return true;
+              if ( paste ) {
```
in which case it's falsy?

charles01pd2018 replied:
this is just to satisfy typescript i don't know to be honest lol

hadielyakhni commented on src/server.test.ts:
```diff
@@ -3,8 +3,10 @@ import { parseMarkdown } from "./server";
-    content: [{ type: "paragraph" }],
-    type: "doc",
-  });
+  if ( ast ) {
```
in which case it's falsy?

charles01pd2018 replied:
this is just to satisfy typescript i don't know to be honest lol

hadielyakhni commented on src/plugins/PasteHandler.ts:
```diff
@@ -34,11 +34,11 @@ export default class PasteHandler extends Extension {
-            //  uncomment to see the error :) this is to be fixed ASAP
-            return replaceHeaderByStrong(html);
-          },
+          // transformPastedHTML: (html: string) => {
+          //   // TODO: this is a temp fix for pasting headers..
+          //   // uncomment to see the error :) this is to be fixed ASAP
+          //   return replaceHeaderByStrong(html);
+          // },
```
btw, in most cases, all commented code should be removed! If needed later, we can rely on git to get it back

charles01pd2018 replied:
got it will keep in mind for future

hadielyakhni commented on src/domHelpers/index.ts:
```diff
@@ -1,6 +1,19 @@
+    if ( node.nodeType === 1 ) {
+      const nextElemText = childNodes[ index + 1 ]?.textContent;
+
+      // this checks to see if the html is residing within a code block
+      // the next block is a new line, that means the html resides within a code block
+      return !nextElemText ? true :
+        !nextElemText.startsWith( '\n' );
+    }
```
can you give more context about the bug and how this change is solving it?

charles01pd2018 replied:
so basically this solves the bug where when sometimes the note messes up after refresh:
https://streamable.com/mbxydm

 I found out that it's because the HTML check parses the doc for anything that looks like or is HTML code and treats it a HTML doc, so for the one pasted code blog post, it had HTML within code blocks messing up the format.

The desired behavior is to have anything with HTML in the document to be parsed as HTML EXCEPT for HTML within code blocks. The above check is kind of a hack I found out after a lot of console.logging -> anything in a code block that is detected as HTML will be followed with a new line (\n). So if that happens, the above code ignores that HTML block within the check.

I think it should be fine because new lines for the editor are backslashes (\), so I THINK it shouldn't mess with any notes saved. For old HTML notes -> i did some testing and i think it should be fine too.

Apandoh reviewed (approved):
Works in terms of fixes - but take a look at Hadi's last comment?
